### PR TITLE
Chiffee makeover and new features

### PR DESCRIPTION
**Pull request by coffeecrank**

### CODE REVIEW

File: admin.py [hunk 1]
```diff
@@ -1,8 +1,25 @@
 from django.contrib import admin
 
-# Register your models here.
```
Comment (gumulka) [inline admin.py]: This should stay in there for future contributions

File: forms.py [hunk 2]
```diff
@@ -1,8 +1,11 @@
 from django import forms
+
 from .models import CATEGORIES
 
+
+# TODO: use forms for rendering!
```
Comment (gumulka) [inline forms.py]: Is this TODO done or still pending?

File: models.py [hunk 3]
```diff
@@ -1,47 +1,79 @@
-from django.db import models
+import secrets
 
-# Create your models here.
```
Comment (gumulka) [inline models.py]: Same as admin.py

File: models.py [hunk 4]
```diff
@@ -1,47 +1,79 @@
...
+        if not Purchase.objects.filter(key=key).exists():
+            break
+
+    return key
 
-CATEGORIES = (
-		('D', 'Trinken'),
-		('F', 'Snacks'),
-		('I', 'Eis'),
-	)
```
Comment (gumulka) [inline models.py]: Will changing the short identifiers break anything with existing installations?

File: models.py [hunk 5]
```diff
@@ -1,47 +1,79 @@
...
-	product_price = models.FloatField()
-	product_categorie = models.CharField(max_length=1, choices=CATEGORIES, default="D")
-	product_active = models.BooleanField(default=True)
+    class Meta:
+        ordering = ['name']
+
+    name = models.CharField(max_length=200)
+    price = models.FloatField()
+    category = models.IntegerField(choices=CATEGORIES)
+    active = models.BooleanField(default=True)
```
Comment (gumulka) [inline models.py]: Will changing the name break existing models? Or can it be migrated?

File: models.py [hunk 6]
```diff
@@ -1,47 +1,79 @@
...
 
-	def __str__(self):              # __unicode__ on Python 2
-		return (str(self.product_name) + " (" + str(self.product_price) + ")")
+    def __str__(self):
+        return f'{self.name}'
```
Comment (gumulka) [inline models.py]: das f{} work with older python versions?

File: templates/chiffee/base.html [hunk 7]
```diff
@@ -1,53 +1,124 @@
...
-		var t = setTimeout(Time, 5000);
-	}
-	</script>
-	</head>
-	<body onload="Time()">
-		<h1 align="center" id="headline">Chiffee</h1>
-	{% block body %}{% endblock %}
-	</body>
+
+<html lang="en">
```
Comment (gumulka) [inline templates/chiffee/base.html]: Change to de, as our default language is german

File: templates/chiffee/base.html [hunk 8]
```diff
@@ -1,53 +1,124 @@
...
+        }
+    </style>
+
+    {% block head %} {% endblock %}
+
+    <script>
+        function time() {
+            let today = new Date();
+            let hours = today.getHours();
+            let minutes = today.getMinutes();
```
Comment (gumulka) [inline templates/chiffee/base.html]: What is the difference between let and var?

File: templates/chiffee/purchase.html [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+{% load chiffee_tags %}
+
+{% block body %}
+    <form action="{% url 'chiffee:make-purchases' %}" method="post">
+        {% csrf_token %}
+
+        <p>Wie viele willst du kaufen?</p>
+
+        {% for i in quantity_range %}
+            {% if i == quantity %}
```
Comment (gumulka) [inline templates/chiffee/purchase.html]: Can you put the if statement inside the button and just add the active with it?

File: templates/chiffee/purchase.html [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+                            {{ u.first_name }} {{ u.last_name }}
+                        </button>
+                    {% elif u|has_group:'students' %}
+                        <button class="btn btn-block btn-danger btn-lg btn-menu"
+                                name="confirm"
+                                type="submit"
+                                value="{{ u.username }}">
+                            {{ u.first_name }} {{ u.last_name }}
+                        </button>
+                    {% endif %}
```
Comment (gumulka) [inline templates/chiffee/purchase.html]: These three look very similar, what is the difference? And can it be done without copying?

File: templates/chiffee/redirect.html [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+    </script>
+{% endblock %}
+
+{% block body %}
+    {% if done %}
+        <p>Erledigt! Umleiten...</p>
+    {% else %}
+        <p>Etwas ist schief gelaufen! Umleiten...</p>
+    {% endif %}
+{% endblock %}
```
Comment (gumulka) [inline templates/chiffee/redirect.html]: There should be a bigger difference between success and failure. People will not recognize this and assume everything went as planned.

So maybe make the background red for failures and give a better explanation on what went wrong.

File: templatetags/chiffee_tags.py [hunk 12]
```diff
@@ -0,0 +1,8 @@
+from django import template
+
+register = template.Library()
+
+
+@register.filter(name='has_group')
+def has_group(user, group_name):
+    return user.groups.filter(name=group_name).exists()
```
Comment (gumulka) [inline templatetags/chiffee_tags.py]: What is this for?

Comment (coffeecrank) [reply templatetags/chiffee_tags.py]: It's a custom tag for Django templates, used in making purchases when users are divided into groups.

File: templates/chiffee/purchase.html (hunk 10 above)
Comment (coffeecrank) [reply templates/chiffee/purchase.html]: I'm not sure it's possible without JavaScript. The difference is the button class ('btn-danger' etc.), it colors them differently.

File: forms.py (hunk 2 above)
Comment (coffeecrank) [reply forms.py]: This is maybe for future updates, it's not necessary but adds to the quality of code.

File: admin.py (hunk 1 above)
Comment (coffeecrank) [reply admin.py]: Fixed.

File: templates/chiffee/purchase.html (hunk 10 above)
Comment (gumulka) [reply templates/chiffee/purchase.html]: what about:
```
<button class="btn btn-block btn-lg btn-menu {% if u|has_group:'professors' %}btn-primary{% elif u|has_group:'employees' %}btn-success{% elif u|has_group:'students' %}btn-menu{% endif %}"
                                name="confirm"
                                type="submit"
                                value="{{ u.username }}">
                            {{ u.first_name }} {{ u.last_name }}
                        </button>
```

just with nicer formatting.

File: templates/chiffee/base.html (hunk 8 above)
Comment (gumulka) [reply templates/chiffee/base.html]: okay

File: forms.py (hunk 2 above)
Comment (gumulka) [reply forms.py]: Then create an Issue in github and remove the todo.

File: templates/chiffee/purchase.html (hunk 10 above)
Comment (coffeecrank) [reply templates/chiffee/purchase.html]: Ah yeah, you can put it in tags too. :D Will do in a bit.

Comment (coffeecrank) [reply templates/chiffee/purchase.html]: Fixed.

File: templates/chiffee/redirect.html (hunk 11 above)
Comment (coffeecrank) [reply templates/chiffee/redirect.html]: I added a giant error 404 in red. :P

File: templates/chiffee/purchase.html (hunk 9 above)
Comment (coffeecrank) [reply templates/chiffee/purchase.html]: Done.

File: forms.py (hunk 2 above)
Comment (coffeecrank) [reply forms.py]: I removed a TODO.

File: models.py (hunk 4 above)
Comment (coffeecrank) [reply models.py]: Described in the new migrations push.

File: models.py (hunk 5 above)
Comment (coffeecrank) [reply models.py]: Described in the new migrations push.

**Comment by coffeecrank** [comment]
Gonna create a new one for better tracking.

**Pull request closed without merging**